core/src/autogluon/core/utils/time.py

```diff
@@ -0,0 +1,95 @@
+import logging
+import time
+from typing import Optional
+
+from pandas import DataFrame
+
+logger = logging.getLogger(__name__)
+
+
+def time_func(f, args: list = None, kwargs: dict = None, time_limit: float = 0.2, max_repeats: int = 10) -> float:
+    """
+    Returns the average time taken by `f(*args, **kwargs)`.
+
+    Parameters
+    ----------
+    f : callable
+        Function to be called.
+        Must not alter outer context or otherwise could have negative side effects.
+    args : list, default = None
+        args to f
+    kwargs : dict, default = None
+        kwargs to f
+    time_limit : float, default = 0.2
+        Time limit in seconds to spend repeatedly calling `f`.
+        Once time limit is exceeded, return current estimate.
+    max_repeats : int, default = 10
+        Maximum repeats of calling `f` to obtain a better average time estimate.
+
+    Returns
+    -------
+    Average time taken in seconds by `f(*args, **kwargs)`.
+    """
+    if args is None:
+        args = list()
+    if kwargs is None:
+        kwargs = dict()
+    time_start = time.time()
+    f(*args, **kwargs)
+    time_end = time.time()
+    avg_time = time_end - time_start
+    if max_repeats == 1:
+        return avg_time
+    if time_limit is None or avg_time < (time_limit / max_repeats):
```

Comment:
Should we instead make `time_limit` to be the time spent for 1 repeat? Otherwise, the user needs to manually calculate the total `time_limit`. Plus, user might not know the default `max_repeats` is 10 and assumed the time_limit was for 1 repeat.

Reply:
`time_limit` is to limit the amount of time used by the call to `time_func`. I think this is more important than the user trying to specify a `time_limit` for how long the actual function they want to time takes to execute, that is the reason they are calling `time_func` in the first place. `max_repeats` is largely an implementation detail, it only matters to stabilize the estimate, and generally stabilizing the estimate is only meaningful if the function is fast to execute and thus isn't time consuming to repeatedly estimate. (Plus longer functions are likely less variable in their runtime)